enzyme/Enzyme/MLIR/Implementations/LinalgAutoDiffOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,256 @@
+        op->getLoc(), outputs, inputs, indexingMaps, iteratorTypes);
+
+    int numInputs = inputs.size();
+    auto buildFuncReturnOp = [numInputs](OpBuilder &builder, Location loc,
+                                         SmallVector<Value> retargs) {
+      builder.create<linalg::YieldOp>(
+          loc, ValueRange{retargs}.take_front(numInputs));
+      return;
+    };
+
+    Region *newOpRegion = newOp.getBlock()->getParent();
+    int numInputsNewOp = cast<linalg::GenericOp>(newOp).getInputs().size();
+    Region *adjointRegion = &adjoint.getBodyRegion();
+    int numInputsAdjoint = adjoint.getInputs().size();
+    Location loc = op->getLoc();
+    int numCaches = 0;
+    SmallVector<Value> pushCaches;
+
+    auto hook = [newOpRegion, adjointRegion, loc, &numCaches = numCaches,
+                 numInputsNewOp, numInputsAdjoint,
+                 &pushCaches = pushCaches](Type t) {
+      OpBuilder builder(newOpRegion);
+      Value pushCache = builder.create<enzyme::InitOp>(loc, t);
+      pushCaches.push_back(pushCache);
+      newOpRegion->addArgument(t, loc);
+
+      Value popCache =
+          adjointRegion->insertArgument(numInputsAdjoint + numCaches, t, loc);
+      numCaches++;
+      return std::make_pair(pushCache, popCache);
+    };
+
+    gutils->Logic.differentiate(
+        gutils, *linalgOp.getBlock()->getParent(), adjoint.getBodyRegion(),
+        /*parentRegion=*/false, buildFuncReturnOp, hook);
+
+    // TODO add pushCaches to the yield in newOp
+    auto newOpYield = cast<linalg::YieldOp>(
+        cast<linalg::GenericOp>(newOp).getBodyRegion().front().getTerminator());
+    for (Value pc : pushCaches) {
+      newOpYield.getValuesMutable().append(pc);
+    }
+
+    Block *body = &(adjoint.getBodyRegion().front());
+    auto yieldOp = cast<linalg::YieldOp>(body->getTerminator());
+
+    for (auto opOperand : yieldOp.getOperands()) {
+      body->addArgument(opOperand.getType(), opOperand.getLoc());
+    }
+    OpBuilder builderAdd(yieldOp);
+    for (auto &&[index, value] : llvm::enumerate(yieldOp.getOperands())) {
+      Value arg = body->getArgument(outputs.size() + numCaches + index);
+      auto diffeType = cast<AutoDiffTypeInterface>(arg.getType());
+      Value grad =
+          diffeType.createAddOp(builderAdd, value.getLoc(), arg, value);
+      yieldOp.setOperand(index, grad);
+    }
+
+    auto newIndexingMaps = newOp.getIndexingMapsArray();
+    auto indexingMapsAdjoint = adjoint.getIndexingMapsArray();
+    for (int i = 0; i < numCaches; i++) {
+      Value cacheArg = body->getArgument(outputs.size() + i);
+
+      Type ct = cacheArg.getType();
+      Type type = MemRefType::get(shapes, ct);
+      auto alloc = cacheBuilder.create<memref::AllocOp>(
+          op->getLoc(), type, ValueRange(iterationDomains));
+      Value cache = gutils->initAndPushCache(alloc, cacheBuilder);
+      alloc->setAttr(alloc.getOperandSegmentSizesAttrName(),
+                     cacheBuilder.getDenseI32ArrayAttr({1, 0}));
```

Comment:
I agree there was a bug here. Which highlevel API are you referring to? AllocOP does not seem to be exposing anything for me.

Reply:
Absolutely all operations produced from ODS define getters and mutators for all named arguments. For operands, the getter is `getArgName` (with automatic capitalization from `arg_name`), and the mutator is `getArgNameMutable`. The latter returns a mutable object that can be used to change the operand list. For attributes, the getter is `getAttrName` and the setter is `setAttrName` for `attr_name`. These (de)construct the attribute on-the-fly to get you the underlying value, e.g. an integer for integer attributes. There are also `getAttrNameAttr` and `setAttrNameAttr` that don't do the deconstruction.

In this case, it should be possible to do `alloc.getDynamicSizesMutable().erase(1, alloc.getDynamicSizesMutable().size() - 1)`. 

On the general basis, the code operating on specific ops should (almost) never use `Operation*` API. That is, given an object of concrete `FooOp` type, the code must not do `fooOp->`, only `fooOp.`. This will result in subtle problems when somebody inevitably changes the ODS definition. E.g., they swap the order of arguments in it to follow the alphabetical order, and suddenly `->getOperand(0)` is no longer what it used to be (but not null!) and there is no way to know.

And the code must under no circumstances modify the "internal" attributes such as operand segment sizes.

That being said, it's unclear to me why the code here can't just directly construct the correct op.